src/stream/src/executor/merge.rs

```diff
@@ -168,10 +175,20 @@ impl MergeExecutor {
                             select_all.add_upstreams_from(select_new);
                         }
 
-                        // Remove upstreams.
-                        select_all.remove_upstreams(
-                            &update.removed_upstream_actor_id.iter().copied().collect(),
-                        );
+                        if !update.get_removed_upstream_actor_id().is_empty() {
+                            // Remove upstreams.
+                            select_all.remove_upstreams(
+                                &update.removed_upstream_actor_id.iter().copied().collect(),
+                            );
+
+                            let col_idxes =
+                                select_all.buffered_watermarks.keys().cloned().collect_vec();
+                            for col_idx in col_idxes {
+                                if let Some(watermark) = select_all.check_heap(col_idx) {
+                                    yield Message::Watermark(watermark);
```

Comment:
Why `check_heap` when receiving a barrier? Is there any corner cases that can make things go wrong?

Reply:
> Why `check_heap` when receiving a barrier? Is there any corner cases that can make things go wrong?

If the only upstream that does not have watermark in heap is removed

Reply:
Good catch! Please also add some comments here.

Reply:
> Good catch! Please also add some comments here.

resolved